### PR TITLE
Remove `thiserror` dependency

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: cairo/src/error.rs [hunk 1]
```diff
@@ -249,30 +243,79 @@ impl From<ffi::cairo_status_t> for Error {
...
-    Cairo(#[from] Error),
-    #[error("IO error: {0}")]
-    Io(#[from] io::Error),
+    Cairo(Error),
+    Io(io::Error),
+}
+
+impl std::error::Error for IoError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        #[allow(deprecated)]
```
Comment (sdroege) [inline cairo/src/error.rs:260]: Why?

File: glib/src/boxed_any_object.rs [hunk 2]
```diff
@@ -3,24 +3,56 @@
...
 pub enum BorrowError {
-    #[error("type of the inner value is not as requested")]
     InvalidType,
-    #[error("value is already mutably borrowed")]
-    AlreadyBorrowed(#[from] std::cell::BorrowError),
+    AlreadyBorrowed(std::cell::BorrowError),
 }
-#[derive(thiserror::Error, Debug)]
+
+impl std::error::Error for BorrowError {}
```
Comment (sdroege) [inline glib/src/boxed_any_object.rs:18]: Misses the `source()` impl

File: glib/src/boxed_any_object.rs [hunk 3]
```diff
@@ -3,24 +3,56 @@
...
+#[derive(Debug)]
 pub enum BorrowMutError {
-    #[error("type of the inner value is not as requested")]
     InvalidType,
-    #[error("value is already immutably borrowed")]
-    AlreadyMutBorrowed(#[from] std::cell::BorrowMutError),
+    AlreadyMutBorrowed(std::cell::BorrowMutError),
+}
+
+impl std::error::Error for BorrowMutError {}
```
Comment (sdroege) [inline glib/src/boxed_any_object.rs:41]: Misses the `source()` impl

File: glib/src/gstring.rs [hunk 4]
```diff
@@ -1376,18 +1446,57 @@ impl<T> GStringUtf8Error<T> {
...
-    #[error("unable to convert")]
+    NoTrailingNul(GStringNoTrailingNulError<T>),
+    InteriorNul(GStringInteriorNulError<T>),
+    InvalidUtf8(GStringUtf8Error<T>),
     Unspecified(T),
 }
 
+impl<T> std::error::Error for GStringFromError<T>
+where
+    T: fmt::Debug + fmt::Display,
```
Comment (sdroege) [inline glib/src/gstring.rs:1459]: Not required for `T` based on your implementation

```suggestion
```

File: glib/src/gstring.rs [hunk 5]
```diff
@@ -1354,10 +1414,20 @@ impl<T> GStringInteriorNulError<T> {
...
 pub struct GStringUtf8Error<T>(T, std::str::Utf8Error);
 
+impl<T> std::error::Error for GStringUtf8Error<T> where
+    T: fmt::Debug + fmt::Display
+{
+}
+
+impl<T> fmt::Display for GStringUtf8Error<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.1)
```
Comment (sdroege) [inline glib/src/gstring.rs:1427]:
```suggestion
        fmt::Display::fmt(self.1, fmt),
```

File: glib/src/gstring.rs [hunk 6]
```diff
@@ -1331,10 +1381,20 @@ impl<T> GStringNoTrailingNulError<T> {
...
 pub struct GStringInteriorNulError<T>(T, GStrInteriorNulError);
 
+impl<T> std::error::Error for GStringInteriorNulError<T> where
+    T: fmt::Debug + fmt::Display
+{
+}
+
+impl<T> fmt::Display for GStringInteriorNulError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.1)
```
Comment (sdroege) [inline glib/src/gstring.rs:1394]:
```suggestion
        fmt::Display::fmt(self.1, fmt)
```

File: glib/src/gstring.rs [hunk 7]
```diff
@@ -1331,10 +1381,20 @@ impl<T> GStringNoTrailingNulError<T> {
 /// Error type indicating that a buffer had unexpected nul-bytes.
 ///
 /// `T` is the type of the value the conversion was attempted from.
-#[derive(thiserror::Error, Clone, PartialEq, Eq, Debug)]
-#[error("{1}")]
+#[derive(Clone, PartialEq, Eq, Debug)]
 pub struct GStringInteriorNulError<T>(T, GStrInteriorNulError);
 
+impl<T> std::error::Error for GStringInteriorNulError<T> where
+    T: fmt::Debug + fmt::Display
```
Comment (sdroege) [inline glib/src/gstring.rs:1388]:
```suggestion
```

Not needed AFAICS

File: glib/src/gstring.rs [hunk 8]
```diff
@@ -1354,10 +1414,20 @@ impl<T> GStringInteriorNulError<T> {
 /// Error type indicating that a buffer had invalid UTF-8.
 ///
 /// `T` is the type of the value the conversion was attempted from.
-#[derive(thiserror::Error, Clone, PartialEq, Eq, Debug)]
-#[error("{1}")]
+#[derive(Clone, PartialEq, Eq, Debug)]
 pub struct GStringUtf8Error<T>(T, std::str::Utf8Error);
 
+impl<T> std::error::Error for GStringUtf8Error<T> where
+    T: fmt::Debug + fmt::Display
```
Comment (sdroege) [inline glib/src/gstring.rs:1421]:
```suggestion
```

Not needed AFAICS

File: glib/src/gstring.rs [hunk 9]
```diff
@@ -1314,10 +1354,20 @@ macro_rules! gformat {
 /// Error type indicating that a buffer did not have a trailing nul-byte.
 ///
 /// `T` is the type of the value the conversion was attempted from.
-#[derive(thiserror::Error, Clone, PartialEq, Eq, Debug)]
-#[error("data provided is not nul terminated")]
+#[derive(Clone, PartialEq, Eq, Debug)]
 pub struct GStringNoTrailingNulError<T>(T);
 
+impl<T> std::error::Error for GStringNoTrailingNulError<T> where
+    T: fmt::Debug + fmt::Display
```
Comment (sdroege) [inline glib/src/gstring.rs:1361]:
```suggestion
```

Not needed AFAICS

File: cairo/src/error.rs [hunk 10]
```diff
@@ -1,157 +1,155 @@
...
     FreetypeError,
-    #[error("Win32Gdi Error")]
     #[doc(alias = "STATUS_WIN32_GDI_ERROR")]
     Win32GdiError,
-    #[cfg(feature = "v1_16")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "v1_16")))]
-    #[error("Tag Error")]
-    #[doc(alias = "STATUS_TAG_ERROR")]
-    TagError,
-    #[cfg(feature = "v1_18")]
```
Comment (sdroege) [inline cairo/src/error.rs:140]: Why do you remove this and others?

Comment (GuillaumeGomez) [reply cairo/src/error.rs:140]: Wrong update.

File: glib/src/gstring.rs (hunk 9 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs:1361]: Only debug is required

**Comment by GuillaumeGomez** [comment]
Updated!

**Pull request merged**